hashicorp/terraform-provider-tfe
# Improve discovery and loading of credentials from Terraform config files

## Description

This provider is somewhat special, because the API token it uses is also used
by Terraform CLI itself. We can give a nicer experience on the local CLI by
auto-discovering Terraform's copy of the token (if one isn't provided in the
provider config or the `TFE_TOKEN` environment variable).

However, we weren't quite matching Terraform's credential-finding behavior:

- We were respecting the `TERRAFORM_CONFIG` environment variable for the main
  config file location, but not `TF_CLI_CONFIG_FILE`. (Terraform will use
  either, but prefers the latter since it matches the other environment variable
  names it uses.)
- We weren't ever consulting the credentials file
  (`~/.terraform.d/credentials.tfrc.json`, on unix), which is written to by
  `terraform login` and/or credential helper scripts. The HCL library
  understands the credentials JSON, so you could work around this by pointing
  `TERRAFORM_CONFIG` at the credentials file, but if there was anything else
  important in the config file you also had to set `TF_CLI_CONFIG_FILE` so that
  Terraform (but not the provider) could still find it.

This commit fixes both of those: we'll accept a non-default config file location
via either environment variable (preferring `TF_CLI_CONFIG_FILE`), and we'll
_also_ consult the credentials file if present.

If both sources include a token for the desired hostname, the provider will
follow Terraform's behavior and prefer the token from the CLI config file.

## Testing plan

1.  Create a basic config that configures the `tfe` provider and maybe manages one resource. In the provider config, provide a hostname (if you aren't using app.terraform.io) but not a token. 
2. Ensure that your environment doesn't have any values set for `TFE_TOKEN`, `TF_CLI_CONFIG_FILE`, or `TERRAFORM_CONFIG`.
3. Ensure that neither your `~/.terraformrc` or `~/.terraform.d/credentials.tfrc.json` files include a token for the configured TFE host, to start out with.
4. Then:

Note: for the environment variable columns, "x" means "unset"; read any other value as "points to a loadable .terraformrc config file containing `_____` for the configured hostname".

.terraformrc | credentials json | `TF_CLI_ CONFIG_FILE` | `TERRAFORM_ CONFIG` | **Outcome**
-|-|-|-|-
valid token | no token | x | x | Works (finds default config file)
no token | valid token | x | x | Works (finds credentials file, merges it with config file)
valid token | bad token | x | x | Works (config file beats credentials file)
no token | no token | valid token | x | Works (new env var works)
no token | no token | x | valid token | Works (old env var works)
bad token | no token | valid token | x | Works (env var beats config file)
no token | bad token | valid token | x | Works (env var beats credentials file) 
no token | no token | valid token | bad token  | Works (new env var beats old)
valid token | no token | no token | x | Fails (env var replaces config file instead of merging with it)
no token | valid token | no token | x | Works (credentials file is still merged with config file, even if config file is non-default)


## Output from acceptance tests

See CI output in the checks below.
